README.md
# wxLame [![Actions Status](https://github.com/cfgnunes/wxlame/workflows/build/badge.svg)](https://github.com/cfgnunes/wxlame/actions)

wxLame is a free front-end for the [LAME](https://lame.sourceforge.io) MP3 encoder.

![screenshot](doc/screenshot.png)

## Getting Started

### Install from binaries

The pre-compiled binaries have been tested to work on the following platforms:

- Windows XP, Vista, 7, 8, 10
- Ubuntu 16.04, 18.04, 20.04, 22.04
- Mint 20.2, 20.4

Link to download the pre-compiled binary files: <https://github.com/cfgnunes/wxlame/releases>

On Linux, you can also install them via *Personal Package Archive (PPA)*:

```sh
sudo add-apt-repository -y ppa:cfgnunes/ppa
sudo apt update
sudo apt -y install wxlame
```

### Install from sources

Alternatively to downloading the pre-compiled binaries, you can build the application using the source code.

#### Linux

This section assumes you are using Ubuntu 20.04 (Focal Fossa), but the procedure is similar for other Linux distributions.

The prerequisites to build the source is to install the following packages:

```sh
sudo apt -y install build-essential cmake lame
sudo apt -y install libwxgtk3.0-gtk3-dev
```

For old Linux distributions, the package `libwxgtk3.0-dev` (instead of `libwxgtk3.0-gtk3-dev`) needs to be installed.

Compile the sources and install the binaries, entering the following commands in the terminal:

```sh
mkdir build
cd build
cmake ..
cmake --build .
sudo make install
```

#### Windows

This section assumes Windows 10, but the procedure is similar for other Windows versions. The prerequisites to build the source is to install the following dependencies:

- [MinGW-W64 GCC-8.1.0](https://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win32/Personal%20Builds/mingw-builds/8.1.0/threads-posix/sjlj/i686-8.1.0-release-posix-sjlj-rt_v6-rev0.7z) - i686-posix-sjlj
- [wxWidgets 3.1.3](https://github.com/wxWidgets/wxWidgets/releases/download/v3.1.3/wxWidgets-3.1.3-headers.7z) - header files ('include' folder)
- [wxWidgets 3.1.3](https://github.com/wxWidgets/wxWidgets/releases/download/v3.1.3/wxMSW-3.1.3_gcc810_Dev.7z) - development files for MinGW GCC 8.1.0 ('lib' folder)
- [Gettext](https://mlocati.github.io/articles/gettext-iconv-windows.html) - 32 bit, static
- [CMake](https://cmake.org/download/)

Compile the sources and install with following commands in the terminal:

```sh
mkdir build
cd build
cmake -G "MinGW Makefiles" -DCMAKE_INSTALL_PREFIX:PATH=dist ..
cmake -G "MinGW Makefiles" -DCMAKE_INSTALL_PREFIX:PATH=dist ..
cmake --build . --tar
```

To create an installer on Windows, which is optional, the [Inno Setup](http://www.jrsoftware.org/isinfo.php) must to be downloaded and installed.
To generate the installer setup, use file `installer.iss` in the Inno Setup after build the binaries.

## Contributing

If you spot a bug, or want to improve the code, or even improve the content, you can do the following:

- [Open an issue](https://github.com/cfgnunes/wxlame/issues/new) describing the bug or feature idea;
- Fork the project, make changes, and submit a pull request.

## Versioning

This project uses [SemVer](https://semver.org) for versioning. For the versions available, see the [tags on this repository](https://github.com/cfgnunes/wxlame/tags).

## License

This program is licensed under the terms of the GNU General Public License version 3. Available online under:
<https://www.gnu.org/licenses/gpl-3.0.html>


doc/CHANGELOG
# Changelog

## v4.? (20??-??-??)

- Minor improvements
- Updated to GTK3 port of the wxWidgets

## v3.5.1 (2019-04-01)

- New Windows binaries
- Code refactor in make/build process

## v3.5 (2015-02-19)

- Changed toolbar icons
- Code refactor
- Bug fixes

## v3.4 (2015-02-18)

- Changed toolbar icons
- Changed Settings screen
- Minor improvements
- Code refactor
- Bug fixes

## v3.3 (2015-02-07)

- Code refactor
- Bug fixes

## v3.2 (2015-02-07)

- Added command line arguments to input files and directories
- Bug fixes

## v3.1 (2015-02-06)

- Bug fixes

## v3.0 (2015-02-06)

- Added Czech translation
- Added German translation
- Updated to wxWidgets 3.0 library
- Bug fixes

## v2.4.3 (2013-07-20)

- Fix adding duplicates files in list
- Fix running on Windows XP
- Added file counter in statusbar

## v2.4.2 (2013-05-03)

- Added Russian translation

## v2.4.1 (2013-05-02)

- Added Spanish translation

## v2.4 (2013-04-28)

- Added support for translation
- Added Portuguese translation

## v2.3.1 (2013-04-23)

- Minor improvements
- Bug fixes

## v2.3 (2012-02-24)

- Bug fixes

## v2.2 (2012-02-18)

- Changed progress dialog
- Minor improvements

## v2.1 (2012-02-15)

- Minor improvements

## v2.0 (2012-02-14)

- Support for add folders
- Support for Drag & Drop folders

## v1.4 (2012-02-10)

- Bug fixes

## v1.3 (2012-02-07)

- Bug fixes

## v1.2 (2012-02-03)

- Support for Drag & Drop
- Show version of tool used
- Added icons with more resolution (for linux version)
- Fixed deb packcage (for linux version)

## v1.1 (2011-02-07)

- Logo changed
- Minor changes

## v1.0 (2011-01-06)

- Initial release


src/Constants.h
/*
 * This file is part of the wxLame and licensed under the GNU General Public License, version 3
 * http://www.gnu.org/licenses/gpl-3.0.html
 */

#ifndef CONSTANTS_HPP
#define CONSTANTS_HPP

#include <wx/dir.h>
#include <wx/filename.h>
#include <wx/stdpaths.h>
#include <wx/string.h>

const wxString APP_NAME = _T("wxLame");
const wxString APP_VERSION = _T("3.5");
const wxString APP_NAME_WITH_VERSION = APP_NAME + _T(" v") + APP_VERSION;

const wxString APP_COPYRIGHT = _T("(C) 2011-2019 Cristiano Fraga G. Nunes <cfgnunes@gmail.com>");
const wxString APP_WEBSITE = _T("https://github.com/cfgnunes/wxlame");

const wxString APP_TOOL_EXECUTABLE = _T("lame");
const wxString APP_WILDCARD_EXT = _T(
        "Audio files|*.mp3;*.wav;*.aif;*.aiff;*.ape;*.MP3;*.WAV;*.AIF;*.AIFF;*.APE|MP3 files (*.mp3)|*.mp3;*.MP3|Wave files (*.wav)|*.wav;*.WAV|AIFF files (*.aif;*.aiff)|*.aif;*.aiff;*.AIF;*.AIFF|Monkey's audio files (*.ape)|*.ape;*.APE");
const wxString APP_OPEN_EXT = _T("mp3;wav;aif;aiff;ape");

const int BITRATE_VALUES[] = {8, 16, 24, 32, 40, 48, 56, 64, 80, 96, 112, 128, 144, 160, 192, 224, 256, 320};
const int BITRATE_VALUES_SIZE = 18;

const wxString RESAMPLING_VALUES[] = {_T("8"), _T("11.025"), _T("12"), _T("16"), _T("22.05"), _T("24"), _T("32"),
                                      _T("44.1"), _T("48")};

enum ModeValues {
    MODE_JOINT = 1,
    MODE_SIMPLE,
    MODE_FORCE,
    MODE_DUAL_MONO,
    MODE_MONO
};

// Config Group Strings
const wxString CONFIG_GROUP_SYSTEM = _T("/System/");
const wxString CONFIG_GROUP_GENERAL = _T("/General/");
const wxString CONFIG_GROUP_ADVANCED = _T("/Advanced/");
const wxString CONFIG_GROUP_VBR = _T("/VBR/");
const wxString CONFIG_GROUP_AUDIO = _T("/Audio/");

// Config Strings
const wxString CONFIG_STR_AppVersion = _T("AppVersion");
const wxString CONFIG_STR_LastOpenDir = _T("LastOpenDir");
const wxString CONFIG_STR_Bitrate = _T("Bitrate");
const wxString CONFIG_STR_Mode = _T("Mode");
const wxString CONFIG_STR_EnableOutDir = _T("EnableOutDir");
const wxString CONFIG_STR_OutDir = _T("OutDir");
const wxString CONFIG_STR_DeleteFiles = _T("DeleteFiles");
const wxString CONFIG_STR_EnabledVBR = _T("EnabledVBR");
const wxString CONFIG_STR_MaxBitrate = _T("MaxBitrate");
const wxString CONFIG_STR_DisableVBRTag = _T("DisableVBRTag");
const wxString CONFIG_STR_EnforceMinBitrate = _T("EnforceMinBitrate");
const wxString CONFIG_STR_UseABR = _T("UseABR");
const wxString CONFIG_STR_VBRQuality = _T("VBRQuality");
const wxString CONFIG_STR_AverageBitrateABR = _T("AverageBitrateABR");
const wxString CONFIG_STR_CustomOptionsText = _T("CustomOptionsText");
const wxString CONFIG_STR_Resampling = _T("Resampling");
const wxString CONFIG_STR_HighpassEnabled = _T("HighpassEnabled");
const wxString CONFIG_STR_HighpassFreq = _T("HighpassFreq");
const wxString CONFIG_STR_HighpassWidthEnabled = _T("HighpassWidthEnabled");
const wxString CONFIG_STR_HighpassWidth = _T("HighpassWidth");
const wxString CONFIG_STR_LowpassEnabled = _T("LowpassEnabled");
const wxString CONFIG_STR_LowpassFreq = _T("LowpassFreq");
const wxString CONFIG_STR_LowpassWidthEnabled = _T("LowpassWidthEnabled");
const wxString CONFIG_STR_LowpassWidth = _T("LowpassWidth");
const wxString CONFIG_STR_MarkNonOriginal = _T("MarkNonOriginal");
const wxString CONFIG_STR_MarkCopyright = _T("MarkCopyright");
const wxString CONFIG_STR_Crc = _T("Crc");
const wxString CONFIG_STR_EnforceISO = _T("EnforceISO");
const wxString CONFIG_STR_AlgorithmQualitySel = _T("AlgorithmQualitySel");
const wxString CONFIG_STR_CustomOptions = _T("CustomOptions");

// Default Values
const wxString DEFAULT_VALUE_LastOpenDir = wxFileName::GetHomeDir();
const int DEFAULT_VALUE_Bitrate = 192;
const int DEFAULT_VALUE_Mode = MODE_JOINT;
const bool DEFAULT_VALUE_EnableOutDir = false;
const wxString DEFAULT_VALUE_OutDir = wxFileName::GetHomeDir();
const bool DEFAULT_VALUE_DeleteFiles = false;
const bool DEFAULT_VALUE_EnabledVBR = false;
const int DEFAULT_VALUE_MaxBitrate = 320;
const bool DEFAULT_VALUE_DisableVBRTag = false;
const bool DEFAULT_VALUE_EnforceMinBitrate = false;
const bool DEFAULT_VALUE_UseABR = false;
const int DEFAULT_VALUE_VBRQuality = 4;
const int DEFAULT_VALUE_AverageBitrateABR = 128;
const int DEFAULT_VALUE_Resampling = 0;
const bool DEFAULT_VALUE_HighpassEnabled = false;
const int DEFAULT_VALUE_HighpassFreq = 20000;
const bool DEFAULT_VALUE_HighpassWidthEnabled = false;
const int DEFAULT_VALUE_HighpassWidth = 20000;
const bool DEFAULT_VALUE_LowpassEnabled = false;
const int DEFAULT_VALUE_LowpassFreq = 20000;
const bool DEFAULT_VALUE_LowpassWidthEnabled = false;
const int DEFAULT_VALUE_LowpassWidth = 20000;
const bool DEFAULT_VALUE_MarkNonOriginal = false;
const bool DEFAULT_VALUE_MarkCopyright = false;
const bool DEFAULT_VALUE_Crc = false;
const bool DEFAULT_VALUE_EnforceISO = false;
const int DEFAULT_VALUE_AlgorithmQualitySel = 0;
const bool DEFAULT_VALUE_CustomOptions = false;
const wxString DEFAULT_VALUE_CustomOptionsText = _T("-b 192 --clipdetect");

// ID for title in wxListCtrl
int const ID_LIST_FILE = 0;
int const ID_LIST_FOLDER = 1;
int const ID_LIST_FORMAT = 2;

inline wxString GetResourceDir() {
    wxString executablePath = wxStandardPaths::Get().GetExecutablePath();
    wxFileName executableFilename(executablePath);
    wxString resourceDirName = _T("/resource/");
    wxDir dir;
    wxString resourceDir;

#ifdef __LINUX__
    resourceDir = _T("/usr/share/wxlame") + resourceDirName;
    if (dir.Exists(resourceDir))
        return resourceDir;
#endif

    resourceDir = _T(".") + resourceDirName;
    if (dir.Exists(resourceDir))
        return resourceDir;

    resourceDir = _T("..") + resourceDirName;
    if (dir.Exists(resourceDir))
        return resourceDir;

    return executableFilename.GetPath() + resourceDirName;
}

#endif // CONSTANTS_HPP
